tests/executables_src/testDataValidityPropagation.cc

```diff
@@ -0,0 +1,552 @@
+  // it seems that the exceptionDummy is created only when app starts.
+  auto exceptionDummy =
+      boost::dynamic_pointer_cast<ctk::ExceptionDummy>(app.device1.getDeviceModule().device.getBackend());
+  assert(exceptionDummy);
+  exceptionDummy->setValidity("/dev/i1", ctk::DataValidity::faulty);
+
+  app.m2.o1 = 1;
+  app.m2.o1.write();
+  test.stepApplication();
+  devI1.read();
+  devI2.read();
+  BOOST_CHECK(devI1.dataValidity() == ctk::DataValidity::faulty);
+  BOOST_CHECK(devI2.dataValidity() == ctk::DataValidity::ok);
+}
+
+/*
+ * \anchor testDataValidity_1_1 \ref dataValidity_1_1 "1.1"
+ * In ApplicationCore each variable has a data validiy flag attached to it. DataValidity can be 'ok' or 'faulty'.
+ *
+ * Explicit test does not make sense since this is clear if this suite compiles,
+ * i.e. expressions testmod1.i1.dataValidity() and testmod1.o1.dataValidity();
+ */
+
+/*
+ * \anchor testDataValidity_1_2 \ref dataValidity_1_2 "1.2"
+ * This flag is automatically propagated: If any of the inputs of an ApplicationModule is faulty,
+ * the data validity of the module becomes faulty, which means all outputs of this module will automatically be
+ * flagged as faulty.
+ * Fan-outs might be special cases (see 2.4).
+ *
+ * See \ref testDataValidity_2_3_3
+ */
+
+/**
+ * \anchor testDataValidity_1_3 \ref dataValidity_1_3 "1.3"
+ * If a device is in error state, all variables which are read from it shall be marked as 'faulty'. This flag is then
+ * propagated through all the modules (via 1.2) so it shows up in the control system.
+ */
+BOOST_AUTO_TEST_CASE(testDataValidity_1_3) {
+  // set up an application with a faulty device
+  TestApplication1<TestModule1> app;
+  app.deviceError = true;
+  // testable mode cannot be used here, since it would wait on initial values (which are not provided)
+  ctk::TestFacility test(false);
+
+  auto i1 = test.getScalar<int>("/dev/i1");
+  test.runApplication();
+
+  // i1.read() would block here
+  i1.readLatest();
+
+  BOOST_CHECK(i1.dataValidity() == ctk::DataValidity::faulty);
+
+  // if Application does not shutdown, this could be an applicationcore bug that requires a workaround for this test
+  // redmine issue: #8550
+}
+
+/*
+ * \anchor testDataValidity_1_4 \ref dataValidity_1_4 "1.4"
+ * The user code has the possibility to query the data validity of the module
+ *
+ * No explicit test, if test suite compiles it's ok.
+ */
+
+/*
+ * \anchor testDataValidity_1_5 \ref dataValidity_1_5 "1.5"
+ * The user code has the possibility to set the data validity of the module to 'faulty'. However, the user code cannot
+ * actively set the module to 'ok' if any of the module inputs are 'faulty'.
+ *
+ * No explicit test. The module should use increment/decrement mechanism to set invalid state,
+ * which implies it cannot override faulty state.
+ * BUT it is actually possible to override getDataValidity in the Module, so spec does not hold strictly!
+ *
+ */
+
+struct TestApplication16 : ctk::Application {
```

Comment:
Please add some Doxygen comment before the struct definition, otherwise Doxygen will place the testDataValidity_1_5 comment above as a documentation for this struct... (Maybe a separator line like `/******/` is enough)

Reply:
ok